packages/grid/x-data-grid/src/components/toolbar/GridToolbar.tsx

```diff
@@ -9,32 +10,62 @@ import { GridToolbarDensitySelector } from './GridToolbarDensitySelector';
 import { GridToolbarFilterButton } from './GridToolbarFilterButton';
 import { GridToolbarExport, GridToolbarExportProps } from './GridToolbarExport';
 import { useGridRootProps } from '../../hooks/utils/useGridRootProps';
+import { GridToolbarQuickFilter, GridToolbarQuickFilterProps } from './GridToolbarQuickFilter';
+
+const GridToolbarButtonsWrapper = styled('div', {
+  name: 'MuiDataGrid',
+  slot: 'ToolbarButtonsWrapper',
+  overridesResolver: (props, styles) => styles.toolbarButtonsWrapper,
+})(() => ({
+  marginRight: 'auto',
+}));
 
 export interface GridToolbarProps
   extends GridToolbarContainerProps,
-    Pick<GridToolbarExportProps, 'csvOptions' | 'printOptions'> {}
+    Pick<GridToolbarExportProps, 'csvOptions' | 'printOptions'> {
+  /**
+   * Show the quick filter component
+   * @default false
+   */
+  showQuickFilter?: boolean;
+  /**
+   * props passed to the quick filter component
+   */
+  quickFilterOptions?: GridToolbarQuickFilterProps;
+}
 
 const GridToolbar = React.forwardRef<HTMLDivElement, GridToolbarProps>(function GridToolbar(
   props,
   ref,
 ) {
-  const { className, csvOptions, printOptions, ...other } = props;
+  const {
+    className,
+    csvOptions,
+    printOptions,
+    showQuickFilter = false,
+    quickFilterOptions = {},
+    ...other
+  } = props;
   const rootProps = useGridRootProps();
 
   if (
     rootProps.disableColumnFilter &&
     rootProps.disableColumnSelector &&
-    rootProps.disableDensitySelector
+    rootProps.disableDensitySelector &&
+    !showQuickFilter
   ) {
     return null;
   }
 
   return (
     <GridToolbarContainer ref={ref} {...other}>
-      <GridToolbarColumnsButton />
-      <GridToolbarFilterButton />
-      <GridToolbarDensitySelector />
-      <GridToolbarExport csvOptions={csvOptions} printOptions={printOptions} />
+      <GridToolbarButtonsWrapper>
+        <GridToolbarColumnsButton />
+        <GridToolbarFilterButton />
+        <GridToolbarDensitySelector />
+        <GridToolbarExport csvOptions={csvOptions} printOptions={printOptions} />
+      </GridToolbarButtonsWrapper>
+      {showQuickFilter && <GridToolbarQuickFilter {...quickFilterOptions} />}
```

Comment:
An alternative is to remove `GridToolbarButtonsWrapper`  and use a div with `flex: 1` to fill the empty space. Both works.

```suggestion
      <Box sx={{ flex: 1 }} />
      {showQuickFilter && <GridToolbarQuickFilter {...quickFilterOptions} />}
```

One thing to note. It's trying to keep the search bar inline with the buttons on mobile. I think it could show the buttons on the first line, then the search bar in the next line taking all width.

<img width="400" src="https://user-images.githubusercontent.com/42154031/165374105-a24b8282-c90d-4da1-b2b2-ce671d97e9d1.png" alt="image" style="max-width: 100%;">

Reply:
Yes, that's better with a native 'wrap'
![image](https://user-images.githubusercontent.com/45398769/165793816-16925a88-921b-4e04-b052-03a24ac05d77.png)